Fixes #5498 ServletHolder cleanup

gregw opened the pull request:
Various cleanups for #5498 including:
 + renaming multiple `_servlet` fields in inner classes to avoid confusion
 + better comments in prepare method to describe why it is needed
 + call prepare from Invoker servlet
 + The `_servlet` field is not set until after the servlet is initialized
 + Consistent wrapping of `SingleThreadedWrapper` now in `initServlet`
 + The `getServlet` method now looks the volatile `_servlet` to avoid locking if possible
 + The `handle` method now calls `getServletInstance` as servlet will have been initialized in `prepare`

gregw committed:
Fixes #55498 ServletHolder cleanup

Various cleanups including:
 + renaming multiple `_servlet` fields in inner classes to avoid confusion
 + better comments in prepare method to describe why it is needed
 + call prepare from Invoker servlet
 + The `_servlet` field is not set until after the servlet is initialized
 + Consistent wrapping of `SingleThreadedWrapper` now in `initServlet`
 + The `getServlet` method now looks the volatile `_servlet` to avoid locking if possible
 + The `handle` method now calls `getServletInstance` as servlet will have been initialized in `prepare`

gregw committed:
Fixes #5498 ServletHolder cleanup

Various cleanups including:
 + renaming multiple `_servlet` fields in inner classes to avoid confusion
 + better comments in prepare method to describe why it is needed
 + call prepare from Invoker servlet
 + The `_servlet` field is not set until after the servlet is initialized
 + Consistent wrapping of `SingleThreadedWrapper` now in `initServlet`
 + The `getServlet` method now looks the volatile `_servlet` to avoid locking if possible
 + The `handle` method now calls `getServletInstance` as servlet will have been initialized in `prepare`
 + Found and fixed race with making unavaiable servlet available again

gregw committed:
Merge remote-tracking branch 'origin/jetty-9.4.x' into jetty-9.4.x-5498-ServletHolder-cleanup

gregw committed:
Merge remote-tracking branch 'origin/jetty-9.4.x' into jetty-9.4.x-5498-ServletHolder-cleanup

gregw commented:
@joakime @janbartel @lorban nudge

lorban commented on jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ServletHolder.java:
```diff
@@ -1229,36 +1244,48 @@ public UnavailableServlet(UnavailableException unavailableException, Servlet ser
-                _servlet = servlet;
-                _available = System.nanoTime() + TimeUnit.SECONDS.toNanos(unavailableException.getUnavailableSeconds());
+                _unavailableServlet = servlet;
+                _available.set(System.nanoTime() + TimeUnit.SECONDS.toNanos(unavailableException.getUnavailableSeconds()));
```
This doesn't take into account the potential numerical overflow of the timeout timestamp.

lorban commented on jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ServletHolder.java:
```diff
@@ -1229,36 +1244,48 @@ public UnavailableServlet(UnavailableException unavailableException, Servlet ser
             {
-                synchronized (ServletHolder.this)
+                long available = _available.get();
+                if (available < 0)
```
This doesn't take into account the potential numerical overflow of the timeout timestamp.

lorban commented on jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ServletHolder.java:
```diff
@@ -1229,36 +1244,48 @@ public UnavailableServlet(UnavailableException unavailableException, Servlet ser
+                    return;
+                }
+
+                if (available == 0 || System.nanoTime() < available)
```
This doesn't take into account the potential numerical overflow of the timeout timestamp.

lorban reviewed (changes requested):
Your changes look alright, but I found an unrelated bug caused by a mis-use of `System.nanoTime()`.

gregw committed:
Updates from review:

 + fixed nanotime overflow
 + fixed several compiler warnings/suggestions

janbartel commented on jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ServletHolder.java:
```diff
@@ -1229,36 +1237,52 @@ public UnavailableServlet(UnavailableException unavailableException, Servlet ser
-            else if (System.nanoTime() < _available)
-                ((HttpServletResponse)res).sendError(HttpServletResponse.SC_SERVICE_UNAVAILABLE);
-            else
+            while (true)
```
Can we come up with a solution that removes the while(true) - I think these are best avoided if possible.

gregw replied:
Will do, because I can in this situation.  However in general, the while(true) pattern is the best and simplest approach for atomic state machines.   More often than not, we've been wrong when we remove them... but I'm 99.9% sure in this case that we have it right.

gregw committed:
Updates from review:

 + removed while true from unavailable servlet

janbartel reviewed (approved):
LGTM

janbartel commented on jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ServletHolder.java:
```diff
@@ -557,33 +555,35 @@ private void makeUnavailable(final Throwable e)
+                servlet = newInstance();
+            if (servlet instanceof javax.servlet.SingleThreadModel)
+            {
+                destroyInstance(servlet);
```
Actually, this isn't right:  we should _not_ call destroyInstance() because this will call servlet.destroy(), but at this stage we will _not_ have called servlet.init().

gregw replied:
This has opened a rats nest  of issues!   It appears  that we do call predestroy on instances that have been passed in.  As we have tests for this potentially wrong behaviour, I think I will not fully address this in 9, but review during the merge to 10.

For now, I have split out a predestroy method, that is called on instances that have not yet been fully initialisd.

janbartel reviewed (changes requested):
I'm gonna have to retract my LGTM.

gregw committed:
Updates from review:

 + Do not destroy servlets unless init has been called.
 + Added TODOs about calling predestroy on instances not created by the holder.

gregw committed:
Updates from review:

 + Do not destroy servlets unless init has been called.
 + Added TODOs about calling predestroy on instances not created by the holder.
 + improved dump and toString

janbartel reviewed (approved):
LGTM

joakime commented:
Why are we even letting people pull the `Servlet` instance out of the `ServletHolder`?

I question this entirely.
I'm of the opinion that we should just eliminate `ServletHolder.getServlet()`.

gregw commented:
> I'm of the opinion that we should just eliminate `ServletHolder.getServlet()`.

Well perhaps make it private rather than public, but we still need the basic logic..... but making it private is a bridge too far for 9 I think.
